File: backend/app/schemas/almacenamiento.py
"""Schemas del gestor de almacenamiento."""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


EstadoAlmacenamiento = Literal["correcto", "necesita_atencion", "critico"]


class AlmacenamientoCategoriaResponse(BaseModel):
    """Uso resumido de una categoría de almacenamiento."""

    clave: str
    nombre: str
    bytes_usados: int = 0
    bytes_recuperables: int = 0
    items: int = 0
    estado: EstadoAlmacenamiento = "correcto"
    descripcion: str = ""


class AlmacenamientoEmisorResponse(BaseModel):
    """Uso lógico seguro asociado a un emisor."""

    empresa_id: int
    etiqueta: str
    lotes: int = 0
    filas_persistidas: int = 0
    filas_compactables: int = 0
    comprobantes: int = 0
    bytes_estimados: int = 0
    bytes_recuperables: int = 0


class AlmacenamientoResumenResponse(BaseModel):
    """Resumen general del almacenamiento visible para administradores."""

    generated_at: datetime
    estado: EstadoAlmacenamiento
    total_bytes_usados: int = 0
    total_bytes_recuperables: int = 0
    storage_limit_bytes: int | None = None
    disk_total_bytes: int | None = None
    disk_used_bytes: int | None = None
    disk_free_bytes: int | None = None
    categorias: list[AlmacenamientoCategoriaResponse] = Field(default_factory=list)
    emisores: list[AlmacenamientoEmisorResponse] = Field(default_factory=list)


class AlmacenamientoItemResponse(BaseModel):
    """Elemento seleccionable para exportación o limpieza."""

    id: str
    nombre: str
    categoria: str
    bytes_usados: int = 0
    bytes_recuperables: int = 0
    descripcion: str = ""
    created_at: datetime | None = None


class LoteCompactableResponse(BaseModel):
    """Lote cerrado cuyo detalle de filas puede compactarse."""

    id: int
    empresa_id: int
    etiqueta_emisor: str
    estado: str
    total_filas: int
    total_grupos: int
    filas_persistidas: int
    bytes_recuperables: int
    created_at: datetime
    finished_at: datetime | None = None


class CrearExportacionAlmacenamientoRequest(BaseModel):
    """Selección de recursos para crear un ZIP de resguardo."""

    model_config = ConfigDict(extra="forbid")

    lote_ids: list[int] = Field(default_factory=list)
    log_ids: list[str] = Field(default_factory=list)
    temporal_ids: list[str] = Field(default_factory=list)


class ExportacionAlmacenamientoResponse(BaseModel):
    """Metadatos de una exportación preparada para descargar."""

    token: str
    estado: str
    archivo_nombre: str
    checksum_sha256: str
    size_bytes: int
    created_at: datetime
    downloaded_at: datetime | None = None
    released_at: datetime | None = None
    manifest: dict


class ConfirmarLiberacionRequest(BaseModel):
    """Confirmación manual posterior a la descarga del resguardo."""

    model_config = ConfigDict(extra="forbid")

    confirmacion: str = Field(..., min_length=1, max_length=50)


class ConfirmarDescargaExportacionRequest(BaseModel):
    """Confirmación explícita de que el ZIP llegó al cliente."""

    model_config = ConfigDict(extra="forbid")

    checksum_sha256: str = Field(..., min_length=64, max_length=64)
    download_token: str = Field(..., min_length=16, max_length=128)


class LimpiarArchivosAlmacenamientoRequest(BaseModel):
    """Selección explícita de archivos administrados a limpiar."""

    model_config = ConfigDict(extra="forbid")

    ids: list[str] = Field(..., min_length=1)


class AccionAlmacenamientoResponse(BaseModel):
    """Resultado de una acción de mantenimiento."""

    mensaje: str
    bytes_afectados: int = 0
    items_afectados: int = 0

File: frontend/src/types/almacenamiento.ts
export type EstadoAlmacenamiento = "correcto" | "necesita_atencion" | "critico";

export interface AlmacenamientoCategoria {
  clave: string;
  nombre: string;
  bytes_usados: number;
  bytes_recuperables: number;
  items: number;
  estado: EstadoAlmacenamiento;
  descripcion: string;
}

export interface AlmacenamientoEmisor {
  empresa_id: number;
  etiqueta: string;
  lotes: number;
  filas_persistidas: number;
  filas_compactables: number;
  comprobantes: number;
  bytes_estimados: number;
  bytes_recuperables: number;
}

export interface AlmacenamientoResumen {
  generated_at: string;
  estado: EstadoAlmacenamiento;
  total_bytes_usados: number;
  total_bytes_recuperables: number;
  storage_limit_bytes: number | null;
  disk_total_bytes: number | null;
  disk_used_bytes: number | null;
  disk_free_bytes: number | null;
  categorias: AlmacenamientoCategoria[];
  emisores: AlmacenamientoEmisor[];
}

export interface AlmacenamientoItem {
  id: string;
  nombre: string;
  categoria: string;
  bytes_usados: number;
  bytes_recuperables: number;
  descripcion: string;
  created_at: string | null;
}

export interface LoteCompactable {
  id: number;
  empresa_id: number;
  etiqueta_emisor: string;
  estado: string;
  total_filas: number;
  total_grupos: number;
  filas_persistidas: number;
  bytes_recuperables: number;
  created_at: string;
  finished_at: string | null;
}

export interface CrearExportacionAlmacenamiento {
  lote_ids: number[];
  log_ids: string[];
  temporal_ids: string[];
}

export interface ExportacionAlmacenamiento {
  token: string;
  estado: string;
  archivo_nombre: string;
  checksum_sha256: string;
  size_bytes: number;
  created_at: string;
  downloaded_at: string | null;
  released_at: string | null;
  manifest: Record<string, unknown>;
}

export interface AccionAlmacenamiento {
  mensaje: string;
  bytes_afectados: number;
  items_afectados: number;
}
